### DISCUSSION TITLE
Use 'ColorSchemePre' to dynamically load/add a colorscheme plugin?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#27666] [Q&A] [answered]
I'm developing a plugin: https://github.com/linrongbin16/colorbox.nvim.

For performance reason, I want to dynamically load/add a plugin's to nvim on needs.

For now I'm using 'packadd' to add/load/install a nvim (colorscheme) plugin, during the 'ColorSchemePre' event (just before running a 'colorscheme' command).

But there's error (see below screenshot), shows that failed to find the plugin's lua module.

<img width="1916" alt="image" src="https://github.com/neovim/neovim/assets/6496887/5349fd11-8613-4a96-8315-cf49ed97b15f">

In the `colors/kanagawa.vim` file, it is calling `lua require('kanagawa').load()`, while nvim doesn't found the lua module 'kanagawa', I believe the runtimepath or packadd doesn't take effect.

I registered callback with `ColorSchemePre`, and suppose this event can be trigger just before running the `colorscheme kanagawa` command. Here's the source code of my plugin 'colorbox.nvim':

```lua
 vim.api.nvim_create_autocmd("ColorSchemePre", {
    callback = function(event)
      local logger = logging.get("colorbox") --[[@as commons.logging.Logger]]
      -- logger:debug("|colorbox.setup| ColorSchemePre event:%s", vim.inspect(event))
      local ColorNameToColorSpecsMap = require("colorbox.db").get_color_name_to_color_specs_map()
      local spec = ColorNameToColorSpecsMap[event.match]
      local autoload = string.format("%s/autoload", spec.full_pack_path)
      local lua_module = string.format("%s/lua", spec.full_pack_path)
      vim.opt.runtimepath:append(spec.full_pack_path)
      vim.opt.runtimepath:append(autoload)
      vim.opt.runtimepath:append(lua_module)
      vim.cmd(string.format([[packadd %s]], spec.git_path))
    end,
  })
```

See: https://github.com/linrongbin16/colorbox.nvim/blob/bf28119ac1bf42a13a0e5e3bb3327080ad4dd5f3/lua/colorbox.lua?plain=1#L79

How should I correctly add a lua package before applying a colorscheme?

**Comment by zeertzjq** [comment]
You aren't using `:packadd` correctly. Please read the documentation:
```
:pa[ckadd][!] {name}	Search for an optional plugin directory in 'packpath'
			and source any plugin files found.  The directory must
			match:
				pack/*/opt/{name} ~
```
And you should only add the plugin's directory to `'runtimepath'`, not the `autoload` or `lua` directories.

**Reply by linrongbin16** [reply, marked as answer]
FYI:

After some more investigation, I found the root cause is: I registered another callback (which runs `colorscheme kanagawa` command) in `VimEnter` event.

And I believe inside `VimEnter`, running the `colorscheme kanagawa` will not trigger another `ColorSchemePre` event, so the `packadd` and `runtimepath` (inside the callbacks registered to the `ColorSchemePre`) doesn't take effect (because the code is not even been runned).

So I simply wrapped the callbacks (registered in `VimEnter`) with a `vim.schedule`. So it will delay the `colorscheme kanagawa` command in next loop, thus make sure `ColorSchemePre` can be invoke before actually running the `colorscheme kanagawa` command.

I finally fixed this issue in PR: https://github.com/linrongbin16/colorbox.nvim/pull/199/files#diff-15bdf0f8c574ae088a524a7a36c69b1e66d42385719134a40e9dfc60723fa5eb

BTW:

I think I'm using `packadd` in correct way, I understand the `name` in `pa[ckadd][!] {name}` should be the folder of the plugin (not the full path, or relative path).
